lib/tether/tether.go

```diff
@@ -173,128 +184,235 @@ func (t *tether) cleanup() {
+		id   string
+		path string
+		err  error
+	}
 
-			// check if session has never been started or is configured for restart
-			if proc == nil || session.Restart {
-				if proc == nil {
-					log.Infof("Launching process for session %s", session.ID)
-				} else {
-					session.m.Lock()
-					session.Diagnostics.ResurrectionCount++
-
-					// FIXME: we cannot have this embedded knowledge of the extraconfig encoding pattern, but not
-					// currently sure how to expose it neatly via a utility function
-					extraconfig.EncodeWithPrefix(t.sink, session, fmt.Sprintf("guestinfo.vice..sessions|%s", session.ID))
-					log.Warnf("Re-launching process for session %s (count: %d)", session.ID, session.Diagnostics.ResurrectionCount)
-					session.Cmd = *restartableCmd(&session.Cmd)
-					session.m.Unlock()
-				}
+	// so that we can launch multiple sessions in parallel
+	var wg sync.WaitGroup
+	// to collect the errors back from them
+	resultsCh := make(chan results, len(t.config.Sessions))
 
-				err := t.launch(session)
-				if err != nil {
-					detail := fmt.Sprintf("failed to launch %s for %s: %s", session.Cmd.Path, id, err)
-					log.Error(detail)
+	// process the sessions and launch if needed
+	for id, session := range t.config.Sessions {
+		session.Lock()
 
-					// TODO: check if failure to launch this is fatal to everything in this containerVM
-					// 		for now failure to launch at all is terminal
-					return errors.New(detail)
-				}
+		log.Debugf("Processing config for session %s", id)
+		var proc = session.Cmd.Process
 
-				continue
+		// check if session is alive and well
+		if proc != nil && proc.Signal(syscall.Signal(0)) == nil {
+			log.Debugf("Process for session %s is already running (pid: %d)", id, proc.Pid)
+			session.Unlock()
+			continue
+		}
+
+		// check if session has never been started or is configured for restart
+		if proc == nil || session.Restart {
+			if proc == nil {
+				log.Infof("Launching process for session %s", id)
+			} else {
+				session.Diagnostics.ResurrectionCount++
+
+				// FIXME: we cannot have this embedded knowledge of the extraconfig encoding pattern, but not
+				// currently sure how to expose it neatly via a utility function
+				extraconfig.EncodeWithPrefix(t.sink, session, fmt.Sprintf("guestinfo.vice..sessions|%s", id))
+				log.Warnf("Re-launching process for session %s (count: %d)", id, session.Diagnostics.ResurrectionCount)
+				session.Cmd = *restartableCmd(&session.Cmd)
 			}
+			session.Unlock()
+
+			wg.Add(1)
+			go func(session *SessionConfig) {
+				defer wg.Done()
+				resultsCh <- results{
+					id:   session.ID,
+					path: session.Cmd.Path,
+					err:  t.launch(session),
+				}
+			}(session)
+		}
+	}
+
+	wg.Wait()
+	// close the channel
+	close(resultsCh)
+
+	// iterate over the results
+	for result := range resultsCh {
+		if result.err != nil {
+			detail := fmt.Errorf("failed to launch %s for %s: %s", result.path, result.id, result.err)
+			log.Error(detail)
```

Comment:
We are doing that (logging first and then returning), or did I misunderstood what you are asking?